src/ipred_prepare.rs

```diff
@@ -182,177 +193,131 @@ pub unsafe fn rav1d_prepare_intra_edges<BD: BitDepth>(
             || av1_intra_prediction_edges[mode as usize]
                 .needs
                 .contains(Needs::LEFT)
-                && have_left == 0)
+                && !have_left)
     {
-        if !prefilter_toplevel_sb_edge.is_null() {
-            dst_top = &*prefilter_toplevel_sb_edge.offset((x * 4) as isize) as *const BD::Pixel;
+        let px_have = cmp::min(8 * tw, 4 * (w - x)) as usize;
+        let n = px_have + have_left as usize;
+        if let Some(prefilter_toplevel_sb_edge) = prefilter_toplevel_sb_edge {
+            let offset = (x * 4) as usize - have_left as usize;
+            &prefilter_toplevel_sb_edge[offset..offset + n]
         } else {
-            dst_top = &*dst.offset(-(BD::pxstride(stride as usize) as isize)) as *const BD::Pixel;
+            &dst[..n]
         }
-    }
+    } else {
+        &[]
+    };
+
     if av1_intra_prediction_edges[mode as usize]
         .needs
         .contains(Needs::LEFT)
     {
-        let sz = th << 2;
-        let left: *mut BD::Pixel = &mut *topleft_out.offset(-sz as isize) as *mut BD::Pixel;
-        if have_left != 0 {
-            let px_have = cmp::min(sz, h - y << 2);
-            let mut i = 0;
-            while i < px_have {
-                *left.offset((sz - 1 - i) as isize) =
-                    *dst.offset(BD::pxstride(stride as usize) as isize * i as isize - 1);
-                i += 1;
+        let sz = 4 * th as usize;
+        let left = &mut topleft_out[(topleft_origin - sz)..];
```

Comment:
```suggestion
        let left = &mut topleft_out[topleft_origin - sz..];
```

You don't need parentheses next to `..` since `..` has a really late binding preference.  You would need parentheses to call a method on a range, though, like `(0..n).step_by(step)`.